### PR TITLE
feat: Implement custom function module in milvus expr

### LINKED ISSUE #36559: [Feature]: Implement custom function module in milvus expr
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe.

OSPP 2024 project: https://summer-ospp.ac.cn/org/prodetail/247410235?list=org&navpage=org

@zhagnlu 

At present, the milvus scalar query system has implemented some operator support and has implemented an extensible system framework, but for developers,

If you want to easily and dynamically extend your own defined operators, avoid redundant operator definitions, etc., and still lack a custom operator basic component,

This component mainly implements the registration and query functions for operator functions, and is integrated with the front-end Plan definition. Based on this component, developers can quickly E2E develop operators. This project mainly implements this basic component and extends some operators on it.

### Describe the solution you'd like.

- add CallExpr in planparserv2, plan.proto, `core/src/exec`
- add a global function factory
- implement some custom functions

### Describe an alternate solution.

_No response_

### Anything else? (Additional Context)

_No response_

### PR DESCRIPTION
**Pull request by jiangyinzuo**
OSPP 2024 project: https://summer-ospp.ac.cn/org/prodetail/247410235?list=org&navpage=org

Solutions:

- parser (planparserv2)
    - add CallExpr in planparserv2/Plan.g4
    - update parser_visitor and show_visitor
- grpc protobuf
    - add CallExpr in plan.proto
- execution (`core/src/exec`)
    - add `CallExpr` `ValueExpr` and `ColumnExpr` (both logical and physical) for function call and function parameters
- function factory (`core/src/exec/expression/function`)
    - create a global hashmap when starting milvus (see server.go)
    - the global hashmap stores function signatures and their function pointers, the CallExpr in execution engine can get the function pointer by function signature.
- custom functions
    - empty(string)
    - starts_with(string, string)
- add cpp/go unittests and E2E tests

closes: #36559

### CODE REVIEW

File: internal/core/src/exec/expression/CallExpr.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    }
+    expression::FilterFunctionRegisterKey func_sig{
+        this->expr_->fun_name(), std::move(func_param_type_list)};
+    auto& factory = expression::FunctionFactory::Instance();
+    auto function = factory.GetFilterFunction(func_sig);
+    if (function == nullptr) {
+        throw std::runtime_error("FilterScalarFunction " + func_sig.toString() +
+                                 " not found. " + this->expr_->ToString());
+    }
+    // eval inputs in function
```
Comment (jiangyinzuo) [inline internal/core/src/exec/expression/CallExpr.cpp:47]: Evaluate the input arguments here, not in function

File: internal/core/src/exec/expression/CallExpr.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+               expr_->inputs().size(),
+               inputs_.size());
+    std::vector<DataType> func_param_type_list;
+    for (auto input : this->inputs_) {
+        func_param_type_list.push_back(input->type());
+    }
+    expression::FilterFunctionRegisterKey func_sig{
+        this->expr_->fun_name(), std::move(func_param_type_list)};
+    auto& factory = expression::FunctionFactory::Instance();
+    auto function = factory.GetFilterFunction(func_sig);
```
Comment (jiangyinzuo) [inline internal/core/src/exec/expression/CallExpr.cpp:42]: Call `factory.GetFilterFunction` at logical expr or parser

**Review by jiangyinzuo — COMMENTED** [review]
Some todos

**Comment by jiangyinzuo** [comment]
@zhagnlu

File: internal/querynodev2/server.go [hunk 3]
```diff
@@ -356,6 +357,8 @@ func (node *QueryNode) Init() error {
 			return
 		}
 
+		C.InitExecExpressionFunctionFactory()
```
Comment (zhagnlu) [inline internal/querynodev2/server.go:360]: move this interface to InitSegcore

File: internal/core/src/common/Vector.h [hunk 4]
```diff
@@ -100,28 +112,81 @@ class ColumnVector final : public BaseVector {
...
+        return reinterpret_cast<As*>(values_->Data());
+    }
+
+    bool
+    IsBitmap() const {
+        return is_bitmap_;
     }
 
  private:
+    bool is_bitmap_;
```
Comment (zhagnlu) [inline internal/core/src/common/Vector.h:146]: this is not a good implment, can just using BitmapVector as a sub class of baseVector

File: internal/core/src/common/Vector.h [hunk 5]
```diff
@@ -58,18 +59,35 @@ class BaseVector {
 
 using VectorPtr = std::shared_ptr<BaseVector>;
 
+class SimpleVector : public BaseVector {
```
Comment (zhagnlu) [inline internal/core/src/common/Vector.h:62]: why add this class?

File: internal/core/src/exec/expression/function/FunctionFactory.cpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+#include <mutex>
+#include "exec/expression/function/impl/StringFunctions.h"
+#include "log/Log.h"
+
+namespace milvus {
+namespace exec {
+namespace expression {
+
+std::string
+FilterFunctionRegisterKey::toString() const {
```
Comment (zhagnlu) [inline internal/core/src/exec/expression/function/FunctionFactory.cpp:27]: ToString()

File: internal/core/src/common/Vector.h (hunk 4 above)
Comment (jiangyinzuo) [reply internal/core/src/common/Vector.h:146]: Yes, but I think this PR is large enough. Replace `ColumnVector` with `BitmapVector`  in other Exprs would introduce a large number of code modifications. It's better to implement `BitmapVector` in a new PR.

File: internal/core/src/common/Vector.h (hunk 5 above)
Comment (jiangyinzuo) [reply internal/core/src/common/Vector.h:62]: In custom functions like `StartsWithVarchar`, we need a base class for `ColumnVector`, `ConstantVector` (and `BitmapVector` in the future). So I add `SimpleVector` as their base class. `BaseVector` only has `size()` and `type` methods, which is unable to meet the requirements.

```cpp
void
StartsWithVarchar(const RowVector& args, FilterFunctionReturn& result) {
    Assert(args.childrens().size() == 2);
    auto strs = std::dynamic_pointer_cast<SimpleVector>(args.child(0));
    if (strs == nullptr) {
        PanicInfo(ExprInvalid, "invalid vector type");
    }
    Assert(strs->type() == DataType::VARCHAR ||
           strs->type() == DataType::STRING);
    auto prefixes = std::dynamic_pointer_cast<SimpleVector>(args.child(1));
    if (prefixes == nullptr) {
        PanicInfo(ExprInvalid, "invalid vector type");
    }
```

File: internal/core/src/common/Vector.h (hunk 4 above)
Comment (jiangyinzuo) [reply internal/core/src/common/Vector.h]: See #37075

**Comment by jiangyinzuo** [comment]
rerun go-sdk

File: tests/python_client/testcases/test_query.py [hunk 7]
```diff
@@ -5598,3 +5598,22 @@ def test_query_text_match_with_unsupported_tokenizer(self):
...
+        expected: query successfully
+        """
+        collection_w, vectors = self.init_collection_general(prefix, insert_data=True,
+                                                             primary_field=ct.default_string_field_name)[0:2]
+        res = vectors[0].iloc[:, 1:3].to_dict('records')
+        output_fields = [default_float_field_name, default_string_field_name]
+        for mixed_call_expr in ["not empty(varchar) && int64 >= 0", "starts_with(varchar, varchar) && int64 >= 0"]:
+            collection_w.query(mixed_call_expr, output_fields=output_fields,
+                               check_task=CheckTasks.check_query_results, check_items={exp_res: res})
+
```
Comment (zhagnlu) [inline tests/python_client/testcases/test_query.py:5619]: add a negative case for not supported function

File: internal/core/src/exec/expression/function/impl/StartsWith.cpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+           prefixes->type() == DataType::STRING);
+
+    TargetBitmap bitmap(strs->size(), false);
+    TargetBitmap valid_bitmap(strs->size(), true);
+    for (size_t i = 0; i < strs->size(); ++i) {
+        if (strs->ValidAt(i) && prefixes->ValidAt(i)) {
+            auto* str_ptr = reinterpret_cast<std::string*>(
+                strs->RawValueAt(i, sizeof(std::string)));
+            auto* prefix_ptr = reinterpret_cast<std::string*>(
+                prefixes->RawValueAt(i, sizeof(std::string)));
```
[the remark this answers is not in the document]
Comment (jiangyinzuo) [reply internal/core/src/exec/expression/function/impl/StartsWith.cpp:53]: `StartsWithVarchar` supports both `starts_with(col1, col2)` and `starts_with(col1, "string literal")`

File: internal/core/src/segcore/SegmentChunkReader.h [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+class SegmentChunkReader {
+ public:
+    SegmentChunkReader(const segcore::SegmentInternalInterface* segment,
+                       int64_t active_count)
+        : segment_(segment), active_count_(active_count) {
+    }
+
+    MultipleChunkDataAccessor
+    GetChunkData(DataType data_type,
```
Comment (czs007) [inline internal/core/src/segcore/SegmentChunkReader.h:47]: Better to rename as GetChunkDataAccessor?

File: internal/core/src/common/Vector.h [hunk 10]
```diff
@@ -100,28 +116,81 @@ class ColumnVector final : public BaseVector {
...
+    bool
+    IsBitmap() const {
+        return is_bitmap_;
     }
 
  private:
+    bool is_bitmap_;  // TODO: remove the field after implementing BitmapVector
     FieldDataPtr values_;
-    FieldDataPtr valid_values_;
+    TargetBitmap valid_values_;  // false means the value is null
```
Comment (czs007) [inline internal/core/src/common/Vector.h:152]: @smellthemoon  please help check this change.

File: internal/core/src/exec/expression/CompareExpr.h [hunk 11]
```diff
@@ -113,31 +100,39 @@ class PhyCompareFilterExpr : public Expr {
...
-          active_count_(active_count),
+          segment_chunk_reader_(segment, active_count),
           batch_size_(batch_size),
           expr_(expr) {
-        is_left_indexed_ = segment_->HasIndex(left_field_);
-        is_right_indexed_ = segment_->HasIndex(right_field_);
-        size_per_chunk_ = segment_->size_per_chunk();
-        if (segment_->is_chunked()) {
+        is_left_indexed_ =
+            segment_chunk_reader_.segment_->HasIndex(left_field_);
```
[the remark this answers is not in the document]
Comment (jiangyinzuo) [reply internal/core/src/exec/expression/CompareExpr.h:107]: Yes, I have extracted some `segment_`-related codes that are commonly used by `CompareExpr` and `ColumnExpr` into the `class SegmentChunkReader`, but there are still some codes in CompareExpr and ColumnExpr that directly use `segment_`.

File: internal/parser/planparserv2/parser_visitor.go [hunk 12]
```diff
@@ -594,6 +594,28 @@ func (v *ParserVisitor) getChildColumnInfo(identifier, child antlr.TerminalNode)
...
+	funcParameters := make([]*planpb.Expr, 0, numParams)
+	for _, param := range ctx.AllExpr() {
+		paramExpr := getExpr(param.Accept(v))
+		funcParameters = append(funcParameters, paramExpr.expr)
+	}
+	return &ExprWithType{
+		expr: &planpb.Expr{
+			Expr: &planpb.Expr_CallExpr{
+				CallExpr: &planpb.CallExpr{
+					FunctionName:       functionName,
```
[the remark this answers is not in the document]
Comment (zhagnlu) [reply internal/parser/planparserv2/parser_visitor.go:610]: > It is auto-generated by protobuf, `UnaryExpr` also uses CamelCase (`Op` and `Child`)
> 
> ```go
> 	if ctx.GetOp() != nil {
> 		expr = &planpb.Expr{
> 			Expr: &planpb.Expr_UnaryExpr{
> 				UnaryExpr: &planpb.UnaryExpr{
> 					Op:    planpb.UnaryExpr_Not,
> 					Child: expr,
> 				},
> 			},
> 		}
> 	}
> ```

I means the function name that input from user

File: internal/core/src/common/Vector.h (hunk 10 above)
Comment (smellthemoon) [reply internal/core/src/common/Vector.h:152]: it's ok to do that, but why not also change the type of values_ to TargetBitmap?

Comment (jiangyinzuo) [reply internal/core/src/common/Vector.h:152]: Currently `values_` in `ColumnVector` may be scalar type vectors or bitmap. We can change the type of values_ to TargetBitmap in `BitmapVector` in the future.

File: internal/parser/planparserv2/parser_visitor.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (jiangyinzuo) [reply internal/parser/planparserv2/parser_visitor.go:610]: OK. Now `functionName` is case-insensitive

File: tests/python_client/testcases/test_query.py (hunk 7 above)
Comment (jiangyinzuo) [reply tests/python_client/testcases/test_query.py:5446]: Done. I also added `EmPty` and `EMPTY` for checking case-insensitive function calls

**Comment by jiangyinzuo** [comment]
rerun cpp-unit-test

**Comment by jiangyinzuo** [comment]
rerun cpp-unit-test

**Review by jiangyinzuo — COMMENTED** [review]
I have extracted another 3 methods from `CompareExpr` and `ColumnExpr`.

File: internal/core/src/segcore/SegmentChunkReader.h [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+                         int64_t& current_chunk_pos) const;
+
+    ChunkDataAccessor
+    GetChunkDataAccessor(DataType data_type,
+                         FieldId field_id,
+                         int chunk_id,
+                         int data_barrier) const;
+
+    void
+    MoveCursorForMultipleChunk(int64_t& current_chunk_id,
```
Comment (jiangyinzuo) [inline internal/core/src/segcore/SegmentChunkReader.h:60]: I have extracted another 3 methods from `ColumnExpr` and `CompareExpr`

**Comment by jiangyinzuo** [comment]
rerun cpp-unit-test

**Comment by jiangyinzuo** [comment]
rerun cpp-unit-test

**Comment by jiangyinzuo** [comment]
rerun cpp-unit-test

**Comment by jiangyinzuo** [comment]
rerun cpp-unit-test

**Pull request merged**